### PR TITLE
LLVM API-based codegen

### PR DESCRIPTION
**Pull request by sbogolepov**

### CODE REVIEW

File: libllvmext/src/main/cpp/CodeGen.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  switch (config.outputKind) {
+    case OUTPUT_KIND_BITCODE:
+      codeGenPasses.add(createBitcodeWriterPass(os));
+      break;
+    case OUTPUT_KIND_OBJECT_FILE:
+      targetMachine->addPassesToEmitFile(codeGenPasses, os, getCodeGenFileType());
+      break;
+  }
+
+
```
Comment (olonho) [inline libllvmext/src/main/cpp/CodeGen.cpp]: Excessive spaces.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/codegen/LlvmLto.kt [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import org.jetbrains.kotlin.backend.konan.reportCompilationError
+import org.jetbrains.kotlin.konan.target.CompilerOutputKind
+import org.jetbrains.kotlin.konan.target.HostManager
+import org.jetbrains.kotlin.konan.target.KonanTarget
+
+internal fun linkBitcode(mainModule: LLVMModuleRef, toLink: List<LLVMModuleRef?>) {
+    for (library in toLink) {
+        val failed = LLVMLinkModules2(mainModule, library)
+        if (failed != 0) {
+            throw Error("failed to link $library") // TODO: retrieve error message from LLVM.
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/codegen/LlvmLto.kt]: This shall be fixed, cryptic/missing error messages isn't very good.

File: libllvmext/build.gradle [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+model {
+    components {
+        llvmext(NativeLibrarySpec) {
+            sources {
+                cpp {
+                    source.srcDirs "src/main/cpp"
+                    exportedHeaders.srcDirs "src/main/include"
+                }
+            }
+            binaries.withType(StaticLibraryBinarySpec) { binary ->
```
Comment (olonho) [inline libllvmext/build.gradle]: Maybe remove it.

File: libllvmext/src/main/cpp/LLVMExt.cpp [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+#include <memory>
+#include <ctime>
+
+using namespace llvm;
+
+class ModuleLinker {
+ public:
+  explicit ModuleLinker(LLVMContext &context)
+      : mergedModule(new Module("merged", context)),
```
Comment (olonho) [inline libllvmext/src/main/cpp/LLVMExt.cpp]: Did you test that this LLVM launcher doesn't leak native memory allocated by LLVM APIs?

File: libllvmext/src/main/cpp/LLVMExt.cpp [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+  std::string cpu = determineTargetCPU(codeGenConfig);
+  std::string targetFeatures = determineTargetFeatures(codeGenConfig);
+  setFunctionAttributes(cpu, targetFeatures, *module);
+
+  CodeGen backend(codeGenConfig);
+  if (backend.compile(std::move(module), output->os())) {
+    return 1;
+  }
+
+//  if (*static_cast<bool *>(context->getDiagnosticContext())) {
```
Comment (olonho) [inline libllvmext/src/main/cpp/LLVMExt.cpp]: Maybe this shall be uncommented?

File: shared/src/main/kotlin/org/jetbrains/kotlin/konan/TempFiles.kt [hunk 6]
```diff
@@ -25,13 +25,15 @@ import org.jetbrains.kotlin.konan.file.*
 class TempFiles(outputPath: String, pathToTemporaryDir: String? = null) {
     private val outputName = File(outputPath).name
 
-    val nativeBinaryFile    by lazy { File(dir,"${outputName}.kt.bc") }
+    val nativeBinaryFile    by lazy { File(dir, "${outputName}.kt.bc") }
```
Comment (olonho) [inline shared/src/main/kotlin/org/jetbrains/kotlin/konan/TempFiles.kt]: Here `$outputName` would do.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/Context.kt [hunk 7]
```diff
@@ -500,11 +503,20 @@ internal class Context(config: KonanConfig) : KonanBackendContext(config) {
 
     fun shouldOptimize() = config.configuration.getBoolean(KonanConfigKeys.OPTIMIZATION)
 
+    fun shouldUseNewBackend(): Boolean {
+        if (config.configuration.getBoolean(KonanConfigKeys.LEGACY_BACKEND)) {
+            return false
+        }
+        // Disable wasm target for now because it needs additional work.
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/Context.kt]: We shall fix it for WASM as well.

**Comment by sbogolepov** [comment]
Obsolete.

**Pull request closed without merging**